File: Data Structure/linked_list/single_ll.js
"use strict";

class Node {
  constructor(data = null, next = null) {
    this.data = data;
    this.next = next;
  }
}

/***
LinkedList Operations:
    1. print(): access each element of the linked list and display them
    2. insertAtBeginning(value): add a new node to the beginning 
    2. insertAtEnd(value): add a new node to the end 
    3. insertAt(value, index): add a new node at a specified index   
    5. removeByValue(value): remove a node 
    6. removeAt(index): remove a node at a specified index   
    7. find(value): find a node 
*/

class SingleLinkedList {
  constructor() {
    this.head = null;
    this.length = 0;
  }

  /**
   * Print the linked list in a concise and clean string representation
   * @param {any types} value
   */
  print() {
    let curr = this.head;
    if (!this.head) {
      console.log("Empty LinkedList");
      return;
    }

    let result = curr.data;
    while (curr.next) {
      curr = curr.next;
      result += ` --> ${curr.data}`;
    }
    console.log(result);
  }

  /**
   * Insert a new value to the beginning of the linked list
   * @param {any types} value
   */
  insertAtBeginning(value) {
    this.length += 1;
    const newNode = new Node(value, this.head);
    this.head = newNode;
  }

  /**
   * Insert a new value to the end of the linked list
   * @param {any types} value
   */
  insertAtEnd(value) {
    this.length += 1;
    const newNode = new Node(value, null);

    // Handle empty linked list
    if (this.head === null) {
      this.head = newNode;
      return;
    }

    // Handle regular case
    let curr = this.head;
    while (curr.next) {
      curr = curr.next;
    }
    curr.next = newNode;
  }

  /**
   * Insert a new value at a specified index to the linked list
   * @param {any types} value
   * @param {Number} index
   */
  insertAt(value, index) {
    if (index > this.length || index < 0)
      throw new Error("Invalid index (out of bound");

    const newNode = new Node(value, null);
    let curr = this.head;

    if (index === 0) {
      this.insertAtBeginning(value);
      return;
    }

    // We want to stop at the position right before the desired index
    for (let i = 0; i < index - 1; i++) {
      curr = curr.next;
    }

    newNode.next = curr.next;
    curr.next = newNode;
  }

  /**
   * Remove a node with the given value from the linked list
   * @param {any types} value
   * @returns the linked list length if successfully removed the node, -1 otherwise.
   */
  removeByValue(value) {
    let curr = this.head;

    if (curr === null) return -1;

    if (curr.data === value) {
      this.head = curr.next;
      this.length--;
      return this.length;
    }

    // Need to stop at the prev node
    while (curr.next) {
      if (curr.next.data === value) {
        curr.next = curr.next.next;
        this.length--;
        return this.length;
      }

      curr = curr.next;
    }

    return -1;
  }

  /**
   * Remove a node at the given index from the linked list
   * @param {Number} index
   * @returns 0 if successfully removed the node, -1 otherwise.
   */
  removeAt(index) {
    let curr = this.head;

    // Check for valid index
    if (index < 0 || index > this.length)
      throw new Error("Invalid index (out of bound");

    // return -1 if linked list is empty
    if (curr === null) return -1;

    // Handle remove the 1st node
    if (index === 0) {
      this.head = curr.next;
      this.length -= 1;
      return this.length;
    }

    // Handle regular case
    for (let i = 0; i < index - 1; i++) {
      curr = curr.next;
    }

    curr.next = curr.next.next ? curr.next.next : null;
    this.length -= 1;
    return this.length;
  }

  /**
   * Find the provided data in the linked list
   * @param {any types} value
   * @returns A boolean: true if data is found, false otherwise.
   */
  find(value) {
    let curr = this.head;

    // Handle empty case
    if (!curr) return false;

    while (curr.next) {
      if (curr.data === value) {
        return true;
      }
      curr = curr.next;
    }

    return false;
  }
}

/*------- TESTING -----------*/
const mySingleLL = new SingleLinkedList();
mySingleLL.insertAtBeginning(1);
mySingleLL.insertAtBeginning(2);
mySingleLL.insertAtBeginning(3);
mySingleLL.insertAtEnd("end");
mySingleLL.print();
mySingleLL.insertAt("insertAt", 0);
mySingleLL.print();
mySingleLL.removeAt(3);
mySingleLL.print();

File: Data Structure/linked_list/single_ll.py
"""

"""
class Node: 
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None

    def get_length(self): 
        count = 0
        curr_node = self.head 
        while curr_node:
            count += 1
            curr_node = curr_node.next 
        return count

    def insert_at_beginning(self, data):
        node = Node(data, self.head)
        self.head = node 

    def insert_at_end(self, data):
        new_node = Node(data, None)
        if (self.head is None):
            self.head = new_node
            return 

        curr_node = self.head 
        while curr_node.next:
            curr_node = curr_node.next
        curr_node.next = new_node
    
    def insert_at(self, index, data): 
        if index < 0 or index >= self.get_length(): 
            raise Exception("Invalid index") 
        if index == 0: 
            self.insert_at_beginning(data)
            return 
    
        count = 0 
        curr_node = self.head
        while (curr_node): 
            if (count == index - 1): 
                curr_node.next = Node(data , curr_node.next)
                return 
            curr_node = curr_node.next
            count += 1 

    def insert_after_value(self, value, data_to_insert):
        if self.head is None:
            return

        curr_node = self.head 
        while(curr_node):
            if (curr_node.data == value): 
                curr_node.next = Node(data_to_insert, curr_node.next)      
                return 

            curr_node = curr_node.next

    # Create a new linked list for the given list 
    def insert_values(self, data_list): 
        self.head = None
        for data in data_list:
            self.insert_at_end(data)
    
    def remove_by_value(self, data):
        if self.head is None:
            return
            
        if (self.head.data == data): 
            self.head = self.head.next 
            return 

        curr_node = self.head 
        while(curr_node.next):
            if (curr_node.next.data == data): 
                curr_node.next = curr_node.next.next    
                return 

            curr_node = curr_node.next

    def remove_at(self, index): 
        if index < 0 or index >= self.get_length(): 
            raise Exception("Invalid index") 
        if index == 0: 
            self.head = self.head.next
            return 

        count = 0 
        curr_node = self.head 
        while curr_node:
            if (count == index - 1): 
                curr_node.next = curr_node.next.next
                return 

            curr_node = curr_node.next
            count += 1 

    def print(self): 
        if (self.head is None): 
            print('Linked list is empty') 
            return
        
        curr_node = self.head 
        linkedlist_str = ''
        while curr_node: 
            linkedlist_str += str(curr_node.data) + ' --> ' 
            curr_node = curr_node.next
        linkedlist_str += 'None'
        print(linkedlist_str)
            

if __name__ == '__main__':
    # ll = LinkedList();
    # ll.insert_at_beginning(5)
    # ll.insert_at_beginning(89)
    # ll.insert_at_end(99)
    # ll.insert_at_end(59)
    # ll.print()
    
    ll2 = LinkedList() 
    ll2.insert_values(['banana', 'orange', 'coconut', 'blueberry', 'mango'])
    ll2.insert_at(3, 'watermelon')
    ll2.insert_after_value('banana', 'papaya')
    ll2.remove_by_value('mango')
    ll2.print()
    # ll2.remove_at(2)


    # ll3 = LinkedList() 
    # ll3.print()
   